boot/bootutil/src/image_validate.c

```diff
@@ -445,36 +413,67 @@ bootutil_img_validate(struct enc_key_data *enc_state, int image_index,
                   goto out;
               }
 
-              if (sig_found) {
+              if (sig_0_found) {
                   /* We already found a signature, but we should only have one. */
                   rc = -1;
                   goto out;
               }
 
-              sig_found = true;
+              sig_0_found = true;
 
-              memcpy(sig, buf, len);
+              memcpy(sig0, buf, len);
+              break;
+          }
+          case EXPECTED_SIG_1_TLV:
+          {
+              if (!EXPECTED_SIG_LEN(len) || len > sizeof(buf)) {
+                  rc = -1;
+                  goto out;
+              }
+              rc = LOAD_IMAGE_DATA(hdr, fap, off, buf, len);
+              if (rc) {
+                  goto out;
+              }
+
+              if (sig_1_found) {
+                  /* We already found a signature, but we should only have one. */
+                  rc = -1;
+                  goto out;
+              }
+
+              sig_1_found = true;
+
+              memcpy(sig1, buf, len);
               break;
           }
         }
     }
 
-    if (sigmask == 0 || !sig_found) {
+    if (sigmask == 0 || !sig_0_found || !sig_1_found) {
         rc = -1;
         goto out;
     }
 
-    ed25519_public_key pub;
-    if (true != compute_pubkey(BOOTLOADER_KEY_M, BOOTLOADER_KEY_N, BOOTLOADER_KEYS, sigmask, pub))
-    {
+    int sig0_idx = sigmask & (1 << 0) ? 0 : 1;
```

Comment:
This sigmask handling is wrong: it assumes that there are exactly two out of the first three bits signed and nothing else.
sigmask is part of the signed data so we really have to honor what the signers are commiting to.

the code as written still works if we verify the assumptions, that is:
* `__builtin_popcount(sigmask) == BOOTLOADER_KEY_M`
* bits above `BOOTLOADER_KEY_N` are zero, so ... `sigmask & ~((1 << BOOTLOADER_KEY_N) - 1) == 0` ? i think

Reply:
(of course we shouldn't use `BOOTLOADER_KEY_M` because we're hardcoding the data structures for exactly two signatures; we should still use `BOOTLOADER_KEY_N` and `_Static_assert(BOOTLOADER_KEY_N >= 2)`)